strawberryfields/engine.py

```diff
@@ -603,21 +604,75 @@ def run_async(self, program: Program, *, compile_options=None, **kwargs) -> Job:
         Keyword Args:
             shots (Optional[int]): The number of shots for which to run the job. If this
                 argument is not provided, the shots are derived from the given ``program``.
+            recompile (bool): specifies if ``program`` should be recompiled
+                using the default compilation options
 
         Returns:
             strawberryfields.api.Job: the created remote job
         """
         # get the specific chip to submit the program to
         compile_options = compile_options or {}
+        recompile = kwargs.pop("recompile", False)
         kwargs.update(self._backend_options)
 
         device = self.device_spec
 
         compiler_name = compile_options.get("compiler", device.default_compiler)
-        msg = f"Compiling program for device {device.target} using compiler {compiler_name}."
-        self.log.info(msg)
 
-        program = program.compile(device=device, **compile_options)
+        program_is_compiled = program.compile_info is not None
+
+        if program_is_compiled and not recompile:
+            # error handling for program compilation:
+            # program was compiled but recompilation was not allowed by the
+            # user
+
+            if (
+                program.compile_info[0].target != device.target
+                or program.compile_info[0]._spec != device._spec
+            ):
+                # program was compiled for a different device
+                raise ValueError(
+                    "Cannot use program compiled with "
+                    f"{program._compile_info[0].target} for target {self.target}. "
+                    'Pass the "recompile=True" keyword argument '
+                    f"to compile with {compiler_name}."
+                )
+
+            if program.compile_info[1] != compiler_name:
+                # program was compiled for the device, but using a different compiler
```

Comment:
I think this is rather an important point, but `compiler_name`, from looking at the source code, is just the **default** compiler specified for the device. For example, the default compiler for `X8` is `Xunitary`. However, this doesn't mean that a program compiled using `Xcov` can't be executed on X8; this program should execute fine, even though it used a different compiler!

If a program _was_ compiled for the device, but using a different compiler, we shouldn't reject the job (via an exception), nor should we force recompilation. Instead, we should simply verify that the program is valid, by running it through `Xstrict`.

For example, assuming recompile=False and `RemoteEngine("X8")`:

* Provided program compiled for X8 using `Xcov != (compiler_name == Xunitary)`: should still run, simply validate

* Provided program compiled for X12 (a _different_ device): reject

Reply:
So I think we should simply remove this if statement here, and we can simply rely on the validation happening on line 668

Reply:
Thanks so much for catching this @josh146! :100: Modified it accordingly